src/main/cpp/src/parse_uri.cu

```diff
@@ -0,0 +1,366 @@
+#include <cudf/detail/null_mask.hpp>
+#include <cudf/detail/valid_if.cuh>
+#include <cudf/lists/lists_column_device_view.cuh>
+#include <cudf/scalar/scalar_factories.hpp>
+#include <cudf/strings/detail/strings_children.cuh>
+#include <cudf/strings/detail/utilities.cuh>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <memory>
+
+namespace spark_rapids_jni {
+
+using namespace cudf;
+
+namespace detail {
+namespace {
+
+// utility to validate a character is valid in a URI
+constexpr bool is_valid_character(char ch, bool alphanum_only)
+{
+  if (alphanum_only) {
+    if (ch >= '-' && ch <= '9' && ch != '/') return true;  // 0-9 and .-
+    if (ch >= 'A' && ch <= 'Z') return true;               // A-Z
+    if (ch >= 'a' && ch <= 'z') return true;               // a-z
+  } else {
+    if (ch >= '!' && ch <= ';' && ch != '"') return true;  // 0-9 and !#%&'()*+,-./
+    if (ch >= '=' && ch <= 'Z' && ch != '>') return true;  // A-Z and =?@
+    if (ch >= '_' && ch <= 'z' && ch != '`') return true;  // a-z and _
+  }
+  return false;
+}
+
+/**
+ * @brief Count the number of characters of each string after parsing the protocol.
+ *
+ * @tparam num_warps_per_threadblock Number of warps in a threadblock. This template argument must
+ * match the launch configuration, i.e. the kernel must be launched with
+ * `num_warps_per_threadblock * cudf::detail::warp_size` threads per threadblock.
+ * @tparam char_block_size Number of characters which will be loaded into the shared memory at a
+ * time.
+ *
+ * @param in_strings Input string column
+ * @param out_counts Number of characters in each decode URL
+ * @param out_validity Bitmask of validity data, updated in funcion
+ */
+template <size_type num_warps_per_threadblock, size_type char_block_size>
+__global__ void parse_uri_protocol_char_counter(column_device_view const in_strings,
+                                                size_type* const out_counts,
+                                                bitmask_type* out_validity)
+{
+  __shared__ char temporary_buffer[num_warps_per_threadblock][char_block_size];
+  __shared__ typename cub::WarpScan<int8_t>::TempStorage cub_storage[num_warps_per_threadblock];
+  __shared__ bool found_token[num_warps_per_threadblock];
+
+  auto const global_thread_id = cudf::detail::grid_1d::global_thread_id();
+  auto const global_warp_id   = static_cast<size_type>(global_thread_id / cudf::detail::warp_size);
+  auto const local_warp_id    = static_cast<size_type>(threadIdx.x / cudf::detail::warp_size);
+  auto const warp_lane        = static_cast<size_type>(threadIdx.x % cudf::detail::warp_size);
+  auto const nwarps     = static_cast<size_type>(gridDim.x * blockDim.x / cudf::detail::warp_size);
+  char* in_chars_shared = temporary_buffer[local_warp_id];
+
+  // Loop through strings, and assign each string to a warp.
+  for (size_type row_idx = global_warp_id; row_idx < in_strings.size(); row_idx += nwarps) {
```

Comment:
Does this loop have the potential to overflow if in_strings.size is something close to maxint?

Reply:
First blush on this says yes. I will investigate.